src/fides/api/ops/task/graph_task.py

```diff
@@ -765,36 +775,62 @@ def termination_fn(*dependent_values: int) -> Tuple[int, ...]:
                 access_request_data.get(
                     str(k), []
                 ),  # Pass in the results of the access request for this collection
-                *[
+                [
                     access_request_data.get(
                         str(upstream_key), []
                     )  # Additionally pass in the original input data we used for the access request. It's helpful in
                     # cases like the EmailConnector where the access request doesn't actually retrieve data.
                     for upstream_key in t.input_keys
                 ],
+                *_evaluate_erasure_dependencies(t, erasure_end_nodes),
             )
             for k, t in env.items()
         }
-        # terminator function waits for all keys
-        dsk[TERMINATOR_ADDRESS] = (termination_fn, *env.keys())
+
+        # root node returns 0 to be consistent with the output of the other erasure tasks
+        dsk[ROOT_COLLECTION_ADDRESS] = 0
+        # terminator function reads and returns the cached erasure results for the entire erasure traversal
+        dsk[TERMINATOR_ADDRESS] = (termination_fn, *erasure_end_nodes)
         update_erasure_mapping_from_cache(dsk, resources, start_function)
         await fideslog_graph_rerun(
             prepare_rerun_graph_analytics_event(
-                privacy_request, env, end_nodes, resources, ActionType.erasure
+                privacy_request, env, access_end_nodes, resources, ActionType.erasure
             )
         )
-        v = delayed(get(dsk, TERMINATOR_ADDRESS, num_workers=1))
 
-        update_cts: Tuple[int, ...] = v.compute()
-        # we combine the output of the termination function with the input keys to provide
-        # a map of {collection_name: records_updated}:
-        erasure_update_map: Dict[str, int] = dict(
-            zip([str(x) for x in env], update_cts)
-        )
+        # using an existing function from dask.core to detect cycles in the generated graph
+        collection_cycle = getcycle(dsk, None)
+        if collection_cycle:
+            raise TraversalError(
+                f"The values for the `erase_after` fields caused a cycle in the following collections {collection_cycle}"
+            )
 
+        v = delayed(get(dsk, TERMINATOR_ADDRESS, num_workers=1))
+        update_cts: Dict[str, int] = v.compute()
+        erasure_update_map = {
+            collection.value: update_cts.get(collection.value, 0) for collection in env
+        }
         return erasure_update_map
```

Comment:
You should be able to replace L809-813 with `return v.compute()` if we cache that an erasure happened when the erasure request exits early on a node when **there is no primary key** or the **collection is write-only**:

Consider adding something like:
```python
 self.resources.cache_erasure(
                f"{self.key}", 0
            )  # Cache that the erasure was performed in case we need to restart
```

before these two locations:
https://github.com/ethyca/fides/blob/main/src/fides/api/ops/task/graph_task.py#L529
https://github.com/ethyca/fides/blob/main/src/fides/api/ops/task/graph_task.py#L543

**This will also fix a previous bug discovered in testing here**: If an erasure node is skipped because it doesn't have a primary key (your `products` collection), and that node later fails, on restart, we attempt to run that node again because there was no history in the cache.  This creates multiple sets of execution logs for that node:

```
[('products', 'in_processing'), ('products', 'complete'), 
('orders_to_refunds', 'in_processing'), ('orders_to_refunds', 'complete'), 
('refunds_to_orders', 'in_processing'), ('refunds_to_orders', 'error'), 

('products', 'in_processing'), ('products', 'complete'),
 ('refunds_to_orders', 'in_processing'), ('refunds_to_orders', 'complete'),
 ('orders', 'in_processing'), ('orders', 'complete'), 
('refunds', 'in_processing'), ('refunds', 'complete'), 
('labels', 'in_processing'), ('labels', 'complete')]
```

Reply:
You could add tests for this in tests/ops/integration_tests/test_execution.py, which has other end-to-end tests for when we're restarting a graph from failure. It's on the mongo and postgres collections which are easier to mock specific node failures.

Reply:
Thanks for the tip on how to test this!